### PR TITLE
colexec: fix external aggregator fallback and bool agg functions reset

### LINKED ISSUE #59043: roachtest: tpchvec/disk failed
[(roachtest).tpchvec/disk failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=2586231&tab=buildLog) on [master@a786c51627fe66e47b4a4445c67b2a9077ae2a93](https://github.com/cockroachdb/cockroach/commits/a786c51627fe66e47b4a4445c67b2a9077ae2a93):

```
The test failed on branch=master, cloud=gce:
test artifacts and logs in: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/artifacts/tpchvec/disk/run_1
	tpchvec.go:535,tpchvec.go:633,tpchvec.go:658,test_runner.go:760: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/bin/roachprod run teamcity-2586231-1610694347-24-n3cpu4:1 -- ./workload run tpch --concurrency=1 --db=tpch --default-vectorize --max-ops=1 --queries=1 {pgurl:1}: exit status 20
		(1) attached stack trace
		  -- stack trace:
		  | main.execCmdWithBuffer
		  | 	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/cluster.go:565
		  | main.(*cluster).RunWithBuffer
		  | 	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/cluster.go:2326
		  | main.baseTestRun
		  | 	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/tpchvec.go:530
		  | main.runTPCHVec
		  | 	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/tpchvec.go:633
		  | main.registerTPCHVec.func2
		  | 	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/tpchvec.go:658
		  | main.(*testRunner).runTest.func2
		  | 	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/test_runner.go:760
		  | runtime.goexit
		  | 	/usr/local/go/src/runtime/asm_amd64.s:1374
		Wraps: (2) /home/agent/work/.go/src/github.com/cockroachdb/cockroach/bin/roachprod run teamcity-2586231-1610694347-24-n3cpu4:1 -- ./workload run tpch --concurrency=1 --db=tpch --default-vectorize --max-ops=1 --queries=1 {pgurl:1}
		Wraps: (3) exit status 20
		Error types: (1) *withstack.withStack (2) *errutil.withPrefix (3) *exec.ExitError

```

<details><summary>More</summary><p>

Artifacts: [/tpchvec/disk](https://teamcity.cockroachdb.com/viewLog.html?buildId=2586231&tab=artifacts#/tpchvec/disk)
Related:
- #58834 roachtest: tpchvec/disk failed  [C-test-failure](https://api.github.com/repos/cockroachdb/cockroach/labels/C-test-failure) [O-roachtest](https://api.github.com/repos/cockroachdb/cockroach/labels/O-roachtest) [O-robot](https://api.github.com/repos/cockroachdb/cockroach/labels/O-robot) [branch-release-20.1](https://api.github.com/repos/cockroachdb/cockroach/labels/branch-release-20.1) [release-blocker](https://api.github.com/repos/cockroachdb/cockroach/labels/release-blocker)

[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2Atpchvec%2Fdisk.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

### PR DESCRIPTION
**Pull request by yuzefovich**
Previously, `reset` method of the ordered aggregator would always set
the flag to reset the internal batch to `true`. However, that batch is
only allocated when `Next` is called at least once with a non-zero batch
coming from the input, which is not the case when the fallback to the
disk-backed strategy occurs in the external aggregator (there, we call
`reset` before we use the operator every time). This would lead to
a nil pointer exception, and it is now fixed.

Our unit tests didn't catch it because we forgot to set the forced
number of repartitions which is now also fixed.

This also revealed a bug with resetting of `bool_and` and `bool_or`
aggregates - we forgot to reset whether they have seen a non-null value
or not.

Fixes: #59043.

Release note: None (no stable release with these bugs)

### CODE REVIEW

**Review by asubiotto — CHANGES_REQUESTED** [review]
Nice, could you add a small regression test for the `bool_{and,or}` bug?

> Reviewed 8 of 8 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/59055#-:-MROwp2oC-WNSiWN41jU:b-rnx1g4)* status: :shipit: complete! 0 of 0 LGTMs obtained

**Comment by yuzefovich** [comment]
Note the `bool_*` bug was found by already existing unit tests (once `maxForcedRepartitions` is set). Or do you mean adding a small unit test that verifies specifically the contents of the aggregates after `reset`? That seems like an overkill.

**Comment by asubiotto** [comment]
But those tests test mainly correctness in the presence of disk spilling, right? What I mean is maybe a logic test or something simple that would trigger the `bool_` bug specifically so we don't have to rely on the side effects of the disk spilling tests to inform us of a regression.

**Comment by yuzefovich** [comment]
As discussed in standup, `AggregateFunc.Reset` method is currently only used by the ordered aggregator when it is a part of the fallback strategy in the external aggregator, so the existing coverage seems reasonable.

**Comment by yuzefovich** [comment]
I'll probably violate some rule by not waiting for an LGTM on this PR, but I had a verbal approval from Alfonso during the standup, and this PR fixes the last known release blocker from our team, so I'll go ahead and merge as is. If anything comes up, we can fix it up later.

TFTR!

bors r+

**Pull request merged**
